src/autoschedulers/anderson2021/autotune_loop.sh

```diff
@@ -0,0 +1,581 @@
+#!/bin/bash
+
+# Autotune the given generator
+if [ $# -lt 7 -o $# -gt 8 ]; then
+    echo "Usage: $0 /path/to/some.generator generatorname halide_target weights_file halide_build_dir parallelism train_only [generator_args_sets]"
+    exit
+fi
+
+set -eu
+
+AUTOSCHEDULER_SRC_DIR=$(dirname $0)
+SCRIPTS_DIR="${AUTOSCHEDULER_SRC_DIR}/scripts"
+source ${SCRIPTS_DIR}/utils.sh
+
+GENERATOR=${1}
+PIPELINE=${2}
+HL_TARGET=${3}
+START_WEIGHTS_FILE=${4}
+HALIDE_BUILD_DIR=${5}
+PARALLELISM=${6}
+TRAIN_ONLY=${7}
+
+get_halide_src_dir ${AUTOSCHEDULER_SRC_DIR} HALIDE_SRC_DIR
+get_autoscheduler_build_dir ${HALIDE_BUILD_DIR} AUTOSCHEDULER_BUILD_DIR
+get_tools_build_dir ${HALIDE_BUILD_DIR} TOOLS_BUILD_DIR
+
+LEARNING_RATE=${LEARNING_RATE:-0.001}
+
+# Read the generator-arg sets into an array. Each set is delimited
+# by space; multiple values within each set are are delimited with ;
+# e.g. "set1arg1=1;set1arg2=foo set2=bar set3arg1=3.14;set4arg2=42"
+if [ $# -ge 8 ]; then
+    IFS=' ' read -r -a GENERATOR_ARGS_SETS_ARRAY <<< "${8}"
+else
+    declare -a GENERATOR_ARGS_SETS_ARRAY=
+fi
+
+# Ensure the length is at least 1
+if [ ${#GENERATOR_ARGS_SETS_ARRAY[@]} -eq 0 ]; then
+    GENERATOR_ARGS_SETS_ARRAY=( '' )
+fi
+
+COMPILATION_TIMEOUT=600s
+BENCHMARKING_TIMEOUT=10s
+
+if [ -z ${CXX+x} ]; then
+    echo The CXX environment variable must be set. Exiting...
+    exit
+fi
+
+RUNGENMAIN="${TOOLS_BUILD_DIR}/RunGenMain.o"
+if [ ! -f $RUNGENMAIN ]; then
+    echo "RunGenMain.o not found. Exiting..."
+    exit
+fi
+
+echo Training target is: ${HL_TARGET}
+
+if [ -z ${GENERATOR} ]; then
+GENERATOR=./bin/demo.generator
+fi
+
+if [ -z ${PIPELINE} ]; then
+PIPELINE=demo
+fi
+
+SEARCH_SPACE_OPTIONS=${SEARCH_SPACE_OPTIONS:-"1111"}
+
+SAMPLES=${SAMPLES_DIR}
+mkdir -p ${SAMPLES}
+
+WEIGHTS=${SAMPLES}/updated.weights
+if [[ -f ${WEIGHTS} ]]; then
+    echo Using existing weights "${WEIGHTS}"
+else
+    # Only copy over the weights if we don't have any already,
+    # so that restarted jobs can continue from where they left off
+    cp ${START_WEIGHTS_FILE} ${WEIGHTS}
+    echo Copying starting weights from ${START_WEIGHTS_FILE} to ${WEIGHTS}
+fi
+
+# We could add this unconditionally, but it's easier to wade thru
+# results if we only add if needed
+#for F in disable_llvm_loop_opt; do
+    #if [[ ! ${HL_TARGET} =~ .*${F}.* ]]; then
+        #HL_TARGET="${HL_TARGET}-${F}"
+    #fi
+#done
+
+get_num_cpu_cores NUM_CPU_CORES
+echo "Number of CPU cores detected as ${NUM_CPU_CORES}"
+
+# A batch of this many samples is built in parallel, and then
+# benchmarked serially.
+BATCH_SIZE=80
+EPOCHS=200
+NUM_GPUS=$(nvidia-smi --query-gpu=name --format=csv,noheader | wc -l)
```

Comment:
This requires the auto tuning of Nvidia graphic cards specifically, which is probably fine, but I would add an error message in case somebody tries this with any other GPU vendor.

Reply:
Yes, definitely, NVidia is not the only GPU vendor out there :-)